### ISSUE TITLE
JSON.parse() returns any, but should return Object

### ISSUE THREAD

**Issue by inad9300** [Microsoft/TypeScript#15225]
**TypeScript Version:** 2.3.0

**Expected behavior:** `JSON.parse()` should return `Object`, as stated in [MDN](https://developer.mozilla.org/en/docs/Web/JavaScript/Reference/Global_Objects/JSON/parse#Return_value).

**Actual behavior:** `JSON.parse()` returns `any`, essentially getting rid of the useful compile-time type checks.

I realized about this while working on a project with a person new to TypeScript. She made a copy/paste mistake and ended up with a code equivalent to this:
```typescript
function doSomething(data: string): Promise<any> {
    return JSON.parse(data)
}
```
For which TypeScript didn't show any error.

**Comment by mhegazy** [comment]
`Object` in JavaScript is similar to `any`.  `JSON.parse` can return `number` | `string`| `boolean` | `null` | `object` | Array<the previous>. 

The type `any` here is intentional. it is meant to ease the use of the function (otherwise you will need to cast on every use). The assumption here is you will validate the input value against some criteria and that you will cast to the correct type.

**Comment by inad9300** [comment]
If that's the policy, fine then. IMHO if one wants ease of use, will go with JavaScript; TypeScript sounds to me like the option to chose when you want precisely type correctness.

**Comment by mohsen1** [comment]
`JSON.parse("1") === 1` is valid JS code

**Comment by inad9300** [comment]
I know. I understand I was wrong on my initial suggestion, `Object` as the return type is not a valid solution. @mhegazy ideas were closer to a correct answer, although I don't quite get the role of `object` (lowercase "o") on his attempt.

After giving it some thought, I think it is not possible to correctly represent `JSON.parse()`'s return type using TypeScript, because it may well return an `Array` of `Array`s recursively. Nevertheless, anything more accurate than `any` would be very welcomed, and a reasonable number (say, 16) of recursion levels may be represented manually, e.g.

```typescript
type JsonObject =
    (number | string | boolean | null | Object) | Array<
        (number | string | boolean | null | Object) | Array<
            (number | string | boolean | null | Object) | Array<any>>>

function parseOne(): JsonObject {
    return JSON.parse('1')
}

const n: number = <number> parseOne()
```

I wonder what is the approach taken by TypeScript in similar recursive cases (if they exist).

**Comment by gcanti** [comment]
> I think it is not possible to correctly represent JSON.parse()'s return type using TypeScript

Maybe

```ts
export type JSONObject = { [key: string]: JSON }
export interface JSONArray extends Array<JSON> {}
export type JSON = null | string | number | boolean | JSONArray | JSONObject
```

**Comment by inad9300** [comment]
I'm amazed that works, but with `JSONArray` as a type complains about a circular reference.

**Comment by aluanhaddad** [comment]
> I realized about this while working on a project with a person new to TypeScript. She made a copy/paste mistake and ended up with a code equivalent to this:

```ts
function doSomething(data: string): Promise<any> {
    return JSON.parse(data)
}
```
> For which TypeScript didn't show any error.

Indeed, but if you are OK with explicitly returning `Promise<any>` it is an indication and errors are not far off anyway.

**Comment by inad9300** [comment]
I don't get your point, sorry.

**Comment by mhegazy** [comment]
This issue seems to have been already addressed or is unactionable at the moment. I am auto-closing it for now.

**Comment by rsazima** [comment]
Hi, if I understood this issue correctly, it would be the cause for the behaviour below. Is that correct?

```
> interface I { id: number }
> let i: I = { id: 'a' }
⨯ Unable to compile TypeScript
[eval].ts (1,5): Type '{ id: string; }' is not assignable to type 'I'.
  Types of property 'id' are incompatible.
    Type 'string' is not assignable to type 'number'. (2322)
> let i: I = JSON.parse(JSON.stringify({ id: 'a' }))
undefined
> i
{ id: 'a' }
```

**Comment by inad9300** [comment]
Yes.

**Comment by VarshilDoshi** [comment]
is it possible to create a typescript function in json??
if possible plz send me some source or link for that..
